sot/utils/utils.py

```diff
@@ -296,3 +296,222 @@ def hashable(obj):
         return True
     except TypeError as e:
         return False
+
+
+class OrderedSet(Generic[T]):
+    """
+    A set that preserves the order of insertion.
+    """
+
+    _data: dict[T, None]
+
+    def __init__(self, items: Iterable[T] | None = None):
+        """
+        Examples:
+            >>> s = OrderedSet([1, 2, 3])
+            >>> s
+            OrderedSet(1, 2, 3)
+            >>> s = OrderedSet()
+            >>> s
+            OrderedSet()
+        """
+        items_ = list(items) if items is not None else []
+        self._data: dict[T, None] = dict.fromkeys(items_)
```

Comment:
```suggestion
        self._data = dict.fromkeys(items) if items is not None else {}
```
dict.fromkeys看起来应该可以直接支持Iterable[T]，所以看起来似乎没有必要先转换成list